Anima/diffusion/mcm/animaGaussianMCMVariableProjectionCost.cxx

```diff
@@ -165,25 +167,47 @@ GaussianMCMVariableProjectionCost::PrepareDataForLLS()
 
     // Compute predicted signals and jacobian
     m_PredictedSignalAttenuations.set_size(nbValues,numCompartments);
+    m_VNLSignals.SetSize(numCompartments);
+    m_VNLSignals.Fill(0.0);
+
+    m_OptimalNNLSWeights.set_size(numCompartments);
+    m_OptimalNNLSWeights.fill(0.0);
+
+    m_CholeskyMatrix.set_size(numCompartments,numCompartments);
+    m_CholeskyMatrix.fill(0.0);
+    for (unsigned int i = 0;i < numCompartments;++i)
+        m_CholeskyMatrix(i,i) = 1.0e-6;
+
+    // Scale factor applied to construct Cholesky matrix to avoid colinearities and wrong results
+    double scaleFactor = 1000.0;
 
-    m_VNLSignals.set_size(nbValues);
     for (unsigned int i = 0;i < nbValues;++i)
     {
-        m_VNLSignals[i] = m_ObservedSignals[i];
-
         for (unsigned int j = 0;j < numCompartments;++j)
         {
             unsigned int indexComp = m_IndexesUsefulCompartments[j];
-            m_PredictedSignalAttenuations(i,j) = m_MCMStructure->GetCompartment(indexComp)->GetFourierTransformedDiffusionProfile(m_SmallDelta, m_BigDelta,
-                                                                                                                                  m_GradientStrengths[i], m_Gradients[i]);
+            double predictedSignal = m_MCMStructure->GetCompartment(indexComp)->GetFourierTransformedDiffusionProfile(m_SmallDelta, m_BigDelta, m_GradientStrengths[i], m_Gradients[i]);
+            m_PredictedSignalAttenuations(i,j) = predictedSignal;
+            m_VNLSignals[j] += scaleFactor * predictedSignal * m_ObservedSignals[i];
+            m_CholeskyMatrix(j,j) += scaleFactor * predictedSignal * predictedSignal;
+
+            for (unsigned int k = 0;k < j;++k)
+            {
+                double tmpVal = scaleFactor * predictedSignal * m_PredictedSignalAttenuations(i,k);
+                m_CholeskyMatrix(j,k) += tmpVal;
+                m_CholeskyMatrix(k,j) += tmpVal;
+            }
         }
     }
 
+    m_CholeskySolver.SetInputMatrix(m_CholeskyMatrix);
+    m_CholeskySolver.PerformDecomposition();
+
     bool negativeWeightBounds = m_MCMStructure->GetNegativeWeightBounds();
     if (negativeWeightBounds)
-        m_VNLSignals *= -1.0;
+        m_OptimalNNLSWeights *= -1.0;
 
-    m_OptimalNNLSWeights = vnl_qr <double> (m_PredictedSignalAttenuations).solve(m_VNLSignals);
+    m_OptimalNNLSWeights = m_CholeskySolver.SolveLinearSystem(m_VNLSignals);
```

Comment:
Je pense qu'on peut le faire in-place ici aussi en fait. Je vois pas le risque vu que `m_OptimalNNLSWeights` n'est utilisé que dans cette fonction.

Reply:
C'est une question de vitesse. J'ai testé les deux mais comme dans l'un tu es obligé de recopier les signaux, finalement tu gagnes rien voire tu perds a passer au in place